Led controller

DustinTrumps opened the pull request:
added interfacing for LEDs
-LEDcontroler subsystem added
-interfacing in robotcontaner added

DustinTrumps committed:
I finally made an LED controller branch🤫
-RobotContainer: trigers for led control
-IntakeConstants: constant for can range
-intake: "public" facing for HaveCoral()
-IntakeIOReal: "logic" for HaveCoral()
-LEDcontroler: the interfacing between the roborio and the Arduino

DustinTrumps committed:
to ceep up to date with main

DustinTrumps committed:
I needed to remove a few repetitive lines for the most likely pull request I will do tonight

DustinTrumps committed:
I noticed more unneeded code so... ya  🤫

DustinTrumps committed:
ok now i am done 😭

DustinTrumps committed:
added comments to my code

DustinTrumps committed:
fix for spotless

DustinTrumps committed:
fix for spotless

TheFlameFish commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -225,6 +229,26 @@ private void configureBindings() {
   }
 
+  private boolean IsEndGame() { // Endgame starts when match time is 20 seconds or less
+    return DriverStation.getMatchTime() <= 20 || DriverStation.isAutonomous() == Boolean.FALSE;
```
Did you mean to do `&&`?

DustinTrumps replied:
yes thanks for pointing that out

TheFlameFish commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -225,6 +229,26 @@ private void configureBindings() {
+    new Trigger(() -> intake.getHasGamePiece()).onTrue(Commands.runOnce(() -> LED.LEDDO("o")));
+    // If intake Has a Game Piece and actively shooting set the color to "S" which is "gay
+    // shoot"(Alton named it that, why? idk)
+    new Trigger(() -> controller.rightTrigger(0.25).getAsBoolean() || intake.getHasGamePiece())
```
Did you mean to do &&?

DustinTrumps replied:
yes thanks for pointing that out

TheFlameFish commented on src/main/java/frc/robot/subsystems/LEDcontroler/LEDcontroler.java:
```diff
@@ -0,0 +1,29 @@
+    }
+  }
+
+  public void LEDDO(String what) {
```
Could you add docs to this? If you do it like below, it'll look fancy and show up in intellisense
```java
/** Lorem ipsum dolor sit amet
  * @param what Insert description of this parameter here
  * @return Description of what the method returns (in this case nothing but good to know)
 */

DustinTrumps replied:
do you like this, or should I add more?
```java
/** sets led the color or pattern
  * @param what character so send to the Arduino
 */

TheFlameFish commented on src/main/java/frc/robot/subsystems/LEDcontroler/LEDcontroler.java:
```diff
@@ -0,0 +1,29 @@
+import edu.wpi.first.wpilibj.SerialPort;
+import java.util.Optional;
+
+public class LEDcontroler {
```
Minor typo, "controler" -> "controller"

Also, I'm not sure of the capitalization here. I'd personally go with smth like `LedController` or `LEDController`, but that's purely opinion-based and up to you.

DustinTrumps replied:
I changed all instances to, LedController

TheFlameFish reviewed (commented):
A few minor things

DustinTrumps committed:
ash segested to fix a few things
- || to &&
- added a desctiption to LEDDO()
- fixed a typo and chaged capitalization

DustinTrumps committed:
fixes for spotless,1 day i will remember

nolanhergert commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -225,6 +229,26 @@ private void configureBindings() {
                         })));
   }
 
+  private boolean IsEndGame() { // Endgame starts when match time is 20 seconds or less
```
Ideally if your code is structured well, you should not need comments. The code should explain itself just by reading it like a sentence. Try reading through the next ~15 lines again, and I think you'll see that there's a lot of repeated text in the comments.

I would reserve comments for when you need a longer explanation or something exceptional is occurring and needs emphasis.

nolanhergert commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -225,6 +229,26 @@ private void configureBindings() {
+    new Trigger(() -> DriverStation.isFMSAttached())
+        .onTrue(Commands.runOnce(() -> LED.getalliance()));
+    // If intake Has Game Piece set color to orange
+    new Trigger(() -> intake.getHasGamePiece()).onTrue(Commands.runOnce(() -> LED.LEDDO("o")));
```
One part that makes the code not "read like a sentence" is the "o" character to mean orange or "game piece". We can talk today about how to restructure things so it reads more like:

`LED.LEDDO(LedConstants.GAME_PIECE)`

Plus we might not want it to be orange this year! :)

nolanhergert replied:
I think the implementation would need to look like the constants files. So something like LedConstants.java. Here's an example of another constants file: https://github.com/AEMBOT/FRC_2025/blob/main/src/main/java/frc/robot/constants/DriveConstants.java

nolanhergert reviewed (commented):
Can you check in the Arduino code as well? I think the spot we have it from last year is good: https://github.com/AEMBOT/FRC_2024/tree/main/LED_Controller

DustinTrumps committed:
Nolan segested to fix a few things
-LedConstants, added Constants for leds
-RobotContainer, LedConstants
-LedController, Constants for reading use
-FRC_2025_LED_Controler, the code for the aduino

DustinTrumps committed:
spotlis check. one day i will remember

nolanhergert committed:
Changes from lab

nolanhergert committed:
More changes

nolanhergert committed:
More things to changes

nolanhergert committed:
Suspected fix for not addressing the whole upper strip

nolanhergert commented:
If we do want to implement reef level mapped to colors, maybe it should apply once every second, assuming nothing else is going? I don't know, it's kind of tricky. This also assumes we don't want to show alliance color continuously, which is probably fine.

nolanhergert committed:
Update comment

nolanhergert committed:
Merge remote-tracking branch 'origin/main' into LED-Controller

nolanhergert reviewed (approved):
Good enough for now. All things work as expected, especially that alliance color is set even while disabled and dynamically (can update driver station alliance and it will update right away)

TheFlameFish commented:
> If we do want to implement reef level mapped to colors, maybe it should apply once every second, assuming nothing else is going? I don't know, it's kind of tricky. This also assumes we don't want to show alliance color continuously, which is probably fine.

I think we could probably make it the default command for the subsystem, right?

Ex:
```java
LED.setDefaultCommand(new RepeatCommand(runOnce(() -> {
      // Do reef stuff here
    }).andThen(waitSeconds(0.05))));
```

nolanhergert commented:
> > If we do want to implement reef level mapped to colors, maybe it should apply once every second, assuming nothing else is going? I don't know, it's kind of tricky. This also assumes we don't want to show alliance color continuously, which is probably fine.
> 
> I think we could probably make it the default command for the subsystem, right?
> 
> Ex:
> 
> ```java
> LED.setDefaultCommand(new RepeatCommand(runOnce(() -> {
>       // Do reef stuff here
>     }).andThen(waitSeconds(0.05))));
> ```

I think a default command is the right direction, yes. But we don't want to confuse the driver if, say, a note is intaked and then we basically just blink the rainbow and switch back to reef level right away (which potentially looks like we lost the coral?). I don't know the answer yet, just thinking out loud.

I assume the waitSeconds(0.05) is to not spam the serial right? I think a character every 20ms is probably fine for serial, and I suspect we are also ok on the underlying code too if we repeat the same serial character.
